test/plots/barPlotTests.ts

```diff
@@ -1 +1,783 @@
+          const formatter = (n: number) => `${n}%`;
+          barPlot.labelFormatter(formatter);
+
+          const texts = barPlot.content().selectAll("text");
+          assert.strictEqual(texts.size(), data.length, "one label drawn per datum");
+          const expectedTexts = data.map((d) => formatter(d.value));
+          texts.each(function(d, i) {
+            assert.strictEqual(d3.select(this).text(), expectedTexts[i], `formatter is applied to the displayed value (index ${i})`);
+          });
+        });
+
+        it("shows labels inside or outside the bar as appropriate", () => {
+          barPlot.labelsEnabled(true);
+
+          const labels = barPlot.content().selectAll(".on-bar-label, .off-bar-label");
+          assert.strictEqual(labels.size(), data.length, "one label drawn per datum");
+
+          const bars = barPlot.content().select(".bar-area").selectAll("rect");
+          labels.each((d, i) => {
+            const labelBoundingClientRect = <any> (<SVGElement> labels[0][i]).getBoundingClientRect();
+            const barBoundingClientRect = <any> (<SVGElement> bars[0][i]).getBoundingClientRect();
+            if (labelBoundingClientRect[valueSizeAttr] > barBoundingClientRect[valueSizeAttr]) {
+              assert.isTrue(d3.select(labels[0][i]).classed("off-bar-label"),
+                `label with index ${i} doesn't fit and carries the off-bar class`);
+            } else {
+              assert.isTrue(d3.select(labels[0][i]).classed("on-bar-label"),
+                `label with index ${i} fits and carries the on-bar class`);
+            }
+          });
+        });
+
+        it("shows labels for bars with value = baseline on the \"positive\" side of the baseline", () => {
+          const zeroOnlyData = [ { base: 0, value: 0 } ];
+          dataset.data(zeroOnlyData);
+          barPlot.labelsEnabled(true);
+
+          const label = barPlot.content().select("text");
+          const labelBoundingRect = (<SVGElement> label.node()).getBoundingClientRect();
+          const lineBoundingRect = (<SVGElement> barPlot.content().select(".baseline").node()).getBoundingClientRect();
+          if (isVertical) {
+            const labelPosition = labelBoundingRect.bottom - window.Pixel_CloseTo_Requirement;
+            const linePosition = lineBoundingRect.top;
+            assert.operator(labelPosition, "<=", linePosition, "label with value = baseline is drawn above the baseline");
+          } else {
+            const labelPosition = labelBoundingRect.left + window.Pixel_CloseTo_Requirement;
+            const linePosition = lineBoundingRect.right;
+            assert.operator(labelPosition, ">=", linePosition, "label with value = baseline is drawn to the right of the baseline");
+          }
+        });
+
+        it("hides labels cut off by lower end of base scale", () => {
+          barPlot.labelsEnabled(true);
+          data.forEach((d, i) => {
+            let texts = barPlot.content().selectAll("text");
```

Comment:
Would it be better to have 2 `texts` vars, one for before domain is changed, one for after